### ISSUE TITLE
BatchNormalization fails gradient checking

### ISSUE THREAD

**Issue by ltrottier** [torch/nn#744]
Hi,

I encounter a problem when I was experimenting with optim.checkgrad. When I checked gradInput of BatchNormalization, I got a very large error:

``` lua
require 'nn'
require 'optim'

module = nn.BatchNormalization(5)
input = torch.rand(25,5)
inputSize = input:size()

function feval(x)
    x = x:reshape(inputSize)
    module:forward(x)
    module:backward(x,torch.ones(x:size()))
    return module.output:sum(), module.gradInput:reshape(x:numel())
end

diff, dc, dc_est = optim.checkgrad(feval, input:reshape(input:numel()))
print('Diff must be close to 1e-8: diff = ' .. diff)
```

Output was: 

``` bash
Diff must be close to 1e-8: diff = 1.000000001736
```

When I tried other modules (like Linear, CMul ...), I got around 1e-9. Is there a problem with BatchNormalization or I'm doing something wrong ?

Thanks

**Comment by fmassa** [comment]
BatchNormalization has internal parameters (`running_mean`,  `running_var`, `save_mean`, etc) that changes at every forward, so a standard gradcheck won't work with this module, you will need to fix those variables if you want to do a gradcheck

**Comment by ltrottier** [comment]
Thanks for the answer. However, I was not talking about the gradient of the function with respect to its parameters (gamma and beta), but the gradient of the function with respect to its input. Here, the module takes a NxD matrix as input and outputs a NxD matrix, and because of that, we can only perform gradient checking by using a criterion. In my first post, I used the sum criterion on the output, and my loss function is the simply the criterion itself. To do that, I did not use a criterion explicitly, but instead I backward the 1 matrix NxD and I use the sum of the output in feval. 

So, all parameters that depends on the input should change according to the perturbation on the input. That includes the mean and the variance of the input. That also excludes the running mean and covariance, because they are only used when evaluating the module, not during training.

I personally coded a module that inputs a NxD matrix and outputs a NxD matrix with the mean subtracted. That module passes the gradient checking I used in post one. So if really Batch Normalization should not pass the gradient checking, I don't know what is wrong.

**Comment by ltrottier** [comment]
Sorry, my bad, my module was not subtracting the mean. I checked the math, and I have that the gradient is independent of the input. Maybe that's why gradient checking is not working. Anyway, I'll continue working on that. Thanks!

**Comment by zhangxiaoli73** [comment]
@fmassa in train mode, save_mean is just the sum of input, it should not change at every forward for same input. Is it right? I can't understand why have to fix same_mean for BatchNormalization gradcheck.

**Comment by zhangxiaoli73** [comment]
@ltrottier Hello, have you passed the gradcheck? Can you give me your BatchNormalization gradcheck code? I encounter the same problem. 
Thank you very much. By way,  the input have to subtract mean?
